atc/lidar/scanner.go

```diff
@@ -46,6 +46,17 @@ func (s *scanner) Run(ctx context.Context) error {
 	waitGroup := new(sync.WaitGroup)
 	resourceTypesChecked := &sync.Map{}
 
+	s.scanResources(spanCtx, resources, resourceTypes, waitGroup, resourceTypesChecked)
+	waitGroup.Wait()
```

Comment:
Is there a reason we want the two sets of "scans" to run sequentially? Or can we get rid of this `waitGroup.Wait()`?

Reply:
This gave me a different idea. Got rid of the `sync.Map` and put the `sync.WaitGroup` inside the `scan*` funcs.

Reply:
That's kind of what I had in mind as well, but wasn't totally convinced it would work - e.g. consider the following:

```yaml
resource_types:
- name: grandparent
  type: registry-image
- name: parent
  type: grandparent

resources:
- name: child
  type: parent
```

If we don't perform the recursive `s.check` for parent types, it's possible that we create a check for `parent` before its parent type `grandparent` - might that be an issue? (genuinely not sure, but it feels odd)

Reply:
Didn't even think about that use-case because I don't _think_ someone would use anything else besides `registry-image` as the `type` for `resource_types`. I think this would cause some unexpected behaviour depending on the order of the `resourceTypes` returned by the db. If `grandparent` appears before `parent` then everything is fine. If not, then for **one check interval**, an older version of `grandparent` would be used to fetch `parent`, instead of the most recently available version of `grandparent`. This would immediately be resolved on the next check for these resource-types though. Maybe this is fine? We _eventually_ get you to the expected result? wdyt? I'm a bit on the fence about this.

Reply:
I guess what's not clear to me is what `s.check` actually does. It doesn't actually run the check - it just creates a build in the DB (if it needs to) for the build tracker to pick up in order to actually run the check.

So, I *think* it shouldn't matter - I believe an older version of `grandparent` will always be used to fetch `parent` for a single check interval, even if we create the build for `grandparent` first

Reply:
Good point. I haven't tested this myself but I'm not even sure if there's a max-in-flight for all checks. I remember a PR about smoothing out the number of checks getting run; I forget where that is.
I'm fine with this PR as-is then. Up to you if you think we should still nudge this back to the recursive `s.check` or keep this simpler one.

Reply:
There's some rate limiting for running checks, but that's done within the check step itself (the check delegate, to be more precise):
https://github.com/concourse/concourse/blob/429eead503de082bd1eac1211c7afd0ed20418dc/atc/engine/check_delegate.go#L81

...however, that doesn't seem to rate limit resource types, just resources 🤔

In any event, I think the simplified code is nicer, I'm happy with it 👍